kore/src/Kore/Step/Axiom/EvaluationStrategy.hs

```diff
@@ -92,6 +91,64 @@ definitionEvaluation rules =
     BuiltinAndAxiomSimplifier
         (evaluateWithDefinitionAxioms rules)
 
+{- | Creates an evaluator for a function all the rules that define it.
+
+The function is not applied (@totalDefinitionEvaluation@ returns
+'AttemptedAxiom.NotApplicable') if the supplied rules do not match the entire
+input.
+
+See also: 'definitionEvaluation'
+
+-}
+totalDefinitionEvaluation
+    :: forall level
+    .  [EqualityRule level Variable]
+    -> BuiltinAndAxiomSimplifier level
+totalDefinitionEvaluation rules =
+    BuiltinAndAxiomSimplifier totalDefinitionEvaluationWorker
+  where
+    totalDefinitionEvaluationWorker
+        ::  forall variable
+        .   ( FreshVariable variable
+            , MetaOrObject level
+            , Ord (variable level)
+            , OrdMetaOrObject variable
+            , SortedVariable variable
+            , Show (variable level)
+            , Show (variable Object)
+            , Unparse (variable level)
+            , ShowMetaOrObject variable
+            )
+        => MetadataTools level StepperAttributes
+        -> PredicateSubstitutionSimplifier level
+        -> StepPatternSimplifier level
+        -> BuiltinAndAxiomSimplifierMap level
+        -> StepPattern level variable
+        -> Simplifier
+            ( AttemptedAxiom level variable
+            , SimplificationProof level
+            )
+    totalDefinitionEvaluationWorker
+        tools
+        predicateSimplifier
+        termSimplifier
+        axiomSimplifiers
+        term
+      = do
+        (result, proof) <- evaluate
+        if AttemptedAxiom.hasRemainders result
```

Comment:
Why is it reasonable to not simplify the remainders' predicates before checking whether there are any remainders?

Reply:
The remainders are normalized during rule application by [`Kore.Step.Step.applyRemainder`](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Step.hs#L374), so the predicates should already be simplified.

Reply:
I so wish that we could have a type-safe way (or any way, really) of telling that this is, indeed the case. Preferably a way that's stable through refactoring or other code changes.

Anyway, that's not related to this PR.